File: algo-research/round.py
import random
from math import floor

from seat import Seat
from table import Table

class Round:
    def __init__(self, players=None, tables=None):
        self._tables = tables if tables else []
        self.players = players if players else []

        self.memo_score = None

    def __str__(self):
        value = ""
        for num, table in enumerate(self._tables):
            value += "TABLE #" + str(num) + "\n"
            value += str(table) + "\n"

        return value

    def __repr__(self):
        return str(self)

    def __len__(self):
        return len(self._tables)

    def num_seats(self):
        return sum([len(t) for t in self._tables])

    def clone(self):
        return Round(players=self.players, tables=[t.clone() for t in self._tables])

    def validate(self):
        player_names = [p.name for p in self.players]
        names = self.get_player_names()

        if len(set(names)) != len(player_names):
            raise Exception("Round Validation Failed");

    def should_converge(self, round_num, config):

        # Are the settings such that the tables generates more seats
        # than players to fill them? If so, we will not converge.
        if config.NUM_PLAYERS < self.num_seats():
            return False

        # Are there more players than the number of tables with maximum seating
        # multiplied by the current round number? If so, we will not converge.
        tables_with_maximum_seating = [t for t in self._tables if len(t) == config.SEATS_PER_TABLE]

        if config.NUM_PLAYERS > len(tables_with_maximum_seating) * round_num:
            return False

        # Otherwise, convergence is possible!
        return True

    def score(self, rescore=False):
        if rescore or not self.memo_score:
            self.memo_score = round(sum(map(lambda t: t.score(), self._tables)), 5)
        return self.memo_score

    def meta_score(self):
        seat_scores = sum([t.seat_scores() for t in self._tables], [])
        lowest_seat_score = min([s[1] for s in seat_scores])

        return sum([t.meta_score(lowest_seat_score) for t in self._tables])

    def record_seating(self):
        [t.record_seating() for t in self._tables]

    def get_unlocked_seats(self):
        unlocked_seats = [t.get_unlocked_seats() for t in self._tables]
        return sum(unlocked_seats, [])

    def unlock_seats(self):
        [t.unlock_seats() for t in self._tables]

    def get_player_names(self):
        all_names = set()

        for t in self._tables:
            all_names.update(t.get_player_names())

        return all_names

    def generate_tables(self, num_unseated, max_seats, min_seats):
        self._tables = []

        while num_unseated > 0:
            to_seat = 0

            if num_unseated % max_seats == 0:
                to_seat = max_seats;
            elif num_unseated > min_seats:
                to_seat = min_seats
            else:
                to_seat = num_unseated;

            self._tables.append(Table([Seat() for i in range(0, to_seat)]))

            num_unseated -= to_seat


    def pair_players(self, players, max_table_size, min_table_size):

        self.generate_tables(len(players), max_table_size, min_table_size)

        random.shuffle(players)
        seats = self.get_unlocked_seats()

        while players:
            p = players.pop()
            s = seats.pop()

            s.seat_player(p)


    def _improve(self, config):
        SEAT = 0
        SCORE = 1

        seat_scores = sum([t.seat_scores() for t in self._tables], [])
        lowest_score = min([ss[SCORE] for ss in seat_scores])
        bad_seats = [ss for ss in seat_scores if ss[SCORE] > lowest_score]

        swapped_players = []

        for bad in config.progress(bad_seats):
            better_seats = []

            # Find a better seat for this player
            for t in self._tables:
                test_results = t.test_seating_scores(bad[SEAT].get_player())

                for res in test_results:
                    if res[SCORE] < bad[SCORE]:
                        better_seats.append(res[SEAT])


            if better_seats:
                # Choose randomly a better seat
                new_seat = random.choice(better_seats)
                occupied_player = new_seat.get_player()

                if occupied_player:
                    swapped_players.append(occupied_player)
                    new_seat.unseat_player()

                new_seat.seat_player(bad[SEAT].get_player())
            else:
                swapped_players.append(bad[SEAT].get_player())

            bad[SEAT].unseat_player()

        # shuffle players between all unlocked seats
        new_unlocked_seats = self.get_unlocked_seats()
        random.shuffle(new_unlocked_seats)

        # Randomly seat the swapped out players
        while swapped_players:
            player = swapped_players.pop()
            new_seat = new_unlocked_seats.pop()
            new_seat.seat_player(player)

    def improve(self, config):
        new_round = self.clone()
        new_round._improve(config)

        new_round.unlock_seats()
        new_round.validate()
        return new_round

File: src/app/services/seating/round.js
import Chance from 'chance';
import _ from 'lodash';

import logger from '../../framework/logger';

import Seat from './seat';
import Table from './table';

const chance = new Chance();

export default class Round {

  constructor(players=[], tables=[]) {
    this._tables = tables;
    this.players = players;

    this.memo_score = null;
  }

  to_str() {
    let value = "";

    for(let num in this._tables) {
      const table = this._tables[num];
      value += `TABLE #${num}\n${table.to_str()}\n`;
    }

    return value
  }

  to_repr() {
    return this.to_str();
  }

  get_arrangement() {
    return this._tables.slice(0);
  }

  count() {
    return this._tables.length;
  }

  num_seats() {
    return _.sum(_.map(this._tables, (t) => t.count()));
  }

  clone() {
    const tables = _.map(this._tables, (t) => t.clone());
    return new Round(this.players, tables)
  }

  validate() {
    const all_player_ids = _.map(this.players, (r) => r.get('player_id'));
    const player_ids_at_tables = this.get_player_ids()

    if(_.uniq(player_ids_at_tables).length !== all_player_ids.length) {
      logger.error("Round Validation Failed");
    }
  }

  count_tables_of_size(num_seats) {
    return _.filter(this._tables, (t) => t.count() === num_seats).length;
  }

  should_converge(round_num, config) {
    // No matter what, if this is the first round we should converge on the first
    // attempt at pairing. If we don't something is seriously wrong with the algorithm.
    if(round_num === 1){
      return true;
    }

    if(config.NUM_PLAYERS <= config.SEATS_PER_TABLE) {
      return true;
    }

    // The number of tables > maximum number of seats per table for us to 
    // have a chance to converge.
    if(this.count() < config.SEATS_PER_TABLE) {
      return false;
    }

    // If there are less tables of maximum size than the number of rounds
    // we will not converge
    if(this.count_tables_of_size(config.SEATS_PER_TABLE) <= round_num) {
      return false;
    }

    // Otherwise, convergence is possible!
    return true;
  }

  score(rescore=false) {
    if(rescore || !this.memo_score) {
      const table_scores = _.map(this._tables, (t) => t.score());
      this.memo_score = Math.round(_.sum(table_scores), 5);
    }

    return this.memo_score;
  }

  meta_score() {
    let seat_scores = [];

    for(let t of this._tables) {
      seat_scores = _.concat(seat_scores, t.seat_scores());
    }

    const lowest_seat_score = _.min(_.map(seat_scores, (s) => s[1]));

    return _.sum(_.map(this._tables, (t) => t.meta_score(lowest_seat_score)));
  }

  record_seating() {
    _.map((t) => t.record_seats(), this._tables);
  }

  get_unlocked_seats() {
    const unlocked_seats = _.map(this._tables, (t) => t.get_unlocked_seats());
    return _.reduce(unlocked_seats, (acc, n) => _.concat(acc, n), []);
  }

  unlock_seats() {
    _.map(this._tables, (t) => t.unlock_seats());
  }

  get_player_ids() {
    const ids_at_tables = _.map(this._tables, (t) => t.get_player_ids());
    return _.reduce(ids_at_tables, (acc, n) => _.uniq(_.concat(acc, n)), []);
  }

  generate_tables(num_unseated, max_seats, min_seats) {
    this._tables = []

    while(num_unseated > 0) {
      let to_seat = 0;

      if(num_unseated % max_seats === 0){
        to_seat = max_seats;
      }
      else if(num_unseated > min_seats) {
        to_seat = min_seats;
      }
      else {
        to_seat = num_unseated;
      }

      const new_seats = []
      for(let i=0; i < to_seat; i++) {
        new_seats.push(new Seat());
      }

      this._tables.push(new Table(new_seats));

      num_unseated -= to_seat;
    }
  }

  pair_players(max_table_size, min_table_size) {
    const players = _.shuffle(this.players.slice(0)); //copy the array

    this.generate_tables(players.length, max_table_size, min_table_size)

    const seats = this.get_unlocked_seats()

    while(players.length !== 0) {
      const p = players.pop();
      const s = seats.pop();

      s.seat_player(p);
    }
  }

  _improve(config) {
    const SEAT = 0;
    const SCORE = 1;
    const BETTER_THAN_OCCUPIED_PLAYER = 2;

    const seat_scores = _.reduce(_.map(this._tables, (t) => t.seat_scores()), (acc, n) => _.concat(acc, n), []);
    const lowest_score = _.min(_.map(seat_scores, (ss) => ss[SCORE]));
    const bad_seats = _.filter(seat_scores, (ss) => ss[SCORE] > lowest_score);

    const swapped_players = [];

    for(let bad of bad_seats) {
      const better_seats = [];

      // Find a better seat for this player
      for(let t of this._tables) {
        const test_results = t.test_seating_scores(bad[SEAT].get_player());

        for(let res of test_results) {
          if(res[SCORE] < bad[SCORE]) {// && res[BETTER_THAN_OCCUPIED_PLAYER]) {
            better_seats.push(res);
          }
        }
      }


      if(better_seats.length !== 0) {
        const new_seat = chance.pickone(better_seats)[SEAT];
        const occupied_player = new_seat.get_player();

        if(occupied_player !== null) {
          swapped_players.push(occupied_player);
          new_seat.unseat_player();
        }

        new_seat.seat_player(bad[SEAT].get_player());
      }
      else {
        swapped_players.append(bad[SEAT].get_player());
      }

      bad[SEAT].unseat_player();
    }

    // shuffle players between all unlocked seats
    const new_unlocked_seats = _.shuffle(this.get_unlocked_seats());

    // Randomly seat the swapped out players
    while(swapped_players.length !== 0) {
      const player = swapped_players.pop();
      const new_seat = new_unlocked_seats.pop();
      new_seat.seat_player(player)
    }
  }

  improve(config) {
    const new_round = this.clone();
    new_round._improve(config);

    new_round.unlock_seats();
    new_round.validate();
    return new_round;
  }

}
